### PR TITLE
Sneak highlighting

### LINKED ISSUE #3363: Highlighting (and label mode) in vim-sneak
The easymotion plugin supports highlighting characters, unlike vim-sneak. It'd be helpful to see the next spot one would jump to!

Additionally, label mode is not implemented. However, the argument could be made that we could just use vim-easymotion for the same functionality.

### LINKED ISSUE #4696: <C-o> not working on sneak search
**Describe the bug**
When you do a search with vim-sneak, VSCodeVim doesn't save your position and you can't go back to where you were before using `<C-o>`.

**To Reproduce**

1.  Activate vim-sneak.
2.  Do a forward seach with `s<char><char>`.
3.  Try to go back to where you were before the search using `<C-o>`.
4.  It doesn't go back :(

**Expected behavior**
In normal vim/nvim you can do a search and get back to where you were using `<C-o>`. This is quite common when you need to write a long word which exists somewhere else:

- Jump to that word using sneak.
- Yank it.
- Go back to where you were with `<C-o>`
- Paste the word and continue writing.

**Screenshots**

Doesn't work on VSCodeVim
ezgif com-video-to-gif

But works perfectly on Vim/Nvim
ezgif com-video-to-gif (1)

**Environment (please complete the following information):**

- Extension (VsCodeVim) version: 1.13.1
- VSCode version: 1.43.2
- OS: macOS Catalina

### PR DESCRIPTION
**Pull request by gergelyth**
**What this PR does / why we need it**:

Vim-Sneak highlighting has been missing from the extension and this PR adds that.
It also sets the **isJump** flag, same as #4697 does (although that has tests as well). 

sneak_highlighting

**Which issue(s) this PR fixes**

Fixes #3363
Fixes #4696

### CODE REVIEW

**Comment by gergelyth** [comment]
I'll add label mode as soon as I can.

**Comment by J-Fields** [comment]
Very cool. Is this configurable (either on/off or color) in the original vim sneak? I'd imagine that would be desirable.

**Comment by gergelyth** [comment]
Yes, I added the configuration options already, I just need a few tests for the label mode motion logic and I will push the changes (probably tomorrow).

**Comment by J-Fields** [comment]
Perfect, thanks!

**Comment by gergelyth** [comment]
I managed to finish label mode. I added a few configurations as well so people can customize it.
Previews:
labelmode1
I also implemented the fade-out (as per #4635), because I think it looks awesome.
labelmode2

**Comment by gergelyth** [comment]
I just found 2 bugs, which I haven't considered before:

1. the fade-out decoration only applies to visible ranges, which is a problem when the cursor is at the bottom of the editor and the sneak search jumps to the first target, revealing more text which is not faded-out
2. labels do not get prioritized in the action mapping, meaning that when label mode is active and the markers are showing, if I have a label e.g. `R`, the action performed is the replace instead of the jump

I'll try to fix them as soon as possible.

**Comment by gergelyth** [comment]
I think this is now ready to be merged. I'm not opposed to more testing of course, but I fixed the bugs and I believe the feature is working correctly.

**Comment by gergelyth** [comment]
In the last commit I made a refactor for the sneak label mode to use a different mode for label input: this makes the implementation less error prone and removes the hack from `base.ts`, but has the disadvantage of temporarily removing the selection decoration if sneak is executed in Visual mode (which I guess is preferable - easymotion does this the same way and no one seems to complain about it).

**Comment by joshstobbs** [comment]
Would love to use this, any updates on getting it merged?

**Comment by J-Fields** [comment]
> Would love to use this, any updates on getting it merged?

Just haven't gotten around to it yet, been juggling a bunch of things recently. My expectation is to release a patch update consisting of some bug fixes & performance improvements in the next few days, then land this in the next minor version, hopefully not too long after that.

**Comment by joshstobbs** [comment]
Awesome, thank you for the update!  I'll keep an eye out.

File: src/actions/plugins/sneak.ts [hunk 1]
```diff
@@ -40,98 +113,293 @@ export class SneakForward extends BaseMovement {
...
+  }
 
-    if (this.keysPressed[2] === '\n') {
-      // Single key sneak
-      this.keysPressed[2] = '';
+  protected getLineConditionStopNumber(lineCount: number): number {
+    return lineCount;
+  }
+
+  protected setRepetableMovements(vimState: VimState): VimState {
```
Comment (J-Fields) [inline src/actions/plugins/sneak.ts:259]: typo: repeatable

**Comment by J-Fields** [comment]
Hey @gergelyth would you mind rebasing this and handling the merge conficts?

**Comment by luisherranz** [comment]
I'd love to see this merged as well. I miss sneak highlighting 🙂

**Comment by codeHatcher** [comment]
Would love to see this as well

**Comment by luisherranz** [comment]
@gergelyth can you rebase and solve the merge conflicts? I can't push to this branch but if you can't do it, I can give it a try in a new branch 🙂

**Comment by gergelyth** [comment]
@J-Fields @luisherranz apologies, I've been very busy. I'll get to this sometime this week for sure

**Comment by luisherranz** [comment]
No problem! Thanks @gergelyth :)

**Comment by luisherranz** [comment]
Hey Gergely, any luck? :)

File: src/actions/plugins/sneak.ts (hunk 1 above)
Comment (gergelyth) [reply src/actions/plugins/sneak.ts]: fixed

**Comment by gergelyth** [comment]
Sorry for the delay, I haven't forgotten it, I just had some issues with my local setup.

@J-Fields This is now squashed and rebased, passes both automated and some quick manual tests as well. I believe this is now ready to be merged.

**Comment by J-Fields** [comment]
Ayyy nice! No worries for the delay, I've had very little time for the project recently anyway. Thanks for the work you've put into this - I'm excited about it. I'll try to get around to giving it another read and some manual testing by this weekend.

**Comment by luisherranz** [comment]
Any luck with this @J-Fields? Is there anything I could help with? 🙂

**Comment by joshmu** [comment]
Would love to use this feature!

**Comment by EPonyA** [comment]
What happened to this?

**Comment by joshmu** [comment]
Can we get this over the line? Is there anything I can do to help?

**Comment by gergelyth** [comment]
There are some non-trivial conflicts here I think (just took a quick look), but I can try to rebase again if @J-Fields gives the word.

**Comment by J-Fields** [comment]
Just merged with master - I don't *think* I messed anything up doing so. I tested it for a few minutes and found that with the fade out enabled, it would sometimes fail to cover the first/last line or two. If you get a chance to look at that, @gergelyth, that'd be awesome - otherwise dig in when I get a chance.

**Comment by gergelyth** [comment]
The logic for calculating fade-out ranges is [here](https://github.com/VSCodeVim/Vim/pull/4718/files#diff-a9e393f9576e29fa956f8815c91f365f07b785b55f483f4411ead022f7078107R114-R123).
What it fades out currently:
1. the editor viewport
2. additional lines up until the last match outside of the viewport (+2 lines in the direction of the search for aesthetic reasons, so the last match is prominent as well)

What I believe you saw is - because Sneak jumps automatically to the first result, possibly moving the viewport - some lines not faded out which were originally not in view AND are after the last match (+2 lines). You may also see not faded out lines if you scroll with the mouse (movements such as CTRL+F or CTRL+D will disable sneak).

Now I recognize this may be immersion-breaking / weird / plainly not pretty. I was originally trying to come up with this logic because I was fearing the performance implications of fading out the whole document (especially for large files), but as far as I see we don't have any other choice. Are you aware of such performance issues?
I'll leave it up to you to decide - if we fade out the whole document it should be sufficient to change the highlighted code to:
```typescript
const documentRange = //from first line to last line in the document
editor.setDecorations(this.fadeoutStyle, documentRange);
```

**Comment by sbrocket** [comment]
Apologies if the answer here is obvious, but can you just base the fade-out logic for sneak on easymotion's existing fade-out logic? https://github.com/VSCodeVim/Vim/blob/master/src/actions/plugins/easymotion/easymotion.ts#L224

It'd be awesome to get this landed. :)

**Comment by w-cantin** [comment]
Hi,

I forked this PR at #7162 (I believe I did not have the choice because I do not have the rights to change other people's forks). I think this feature is pretty much ready to merge now and is up to date. I fixed the fadeout issues mentioned by @J-Fields and also improved the sneak label input mode so that the escape key can cancel it at any time. Also action keys are properly triggered now.

I believe fixing the merge conflicts did not mess up anything else as all the test pass correctly.

Let me know if anyone finds any issues or has any ideas on how to improve it.

**Comment by J-Fields** [comment]
Closing this in favor of #7162. Thanks so much @gergelyth for getting this off the ground and dealing with my terrible project management 🙂

**Pull request closed without merging**